Feature #202 xmlparser

chrwm opened the pull request:
Closes https://github.com/OpenEnergyPlatform/open-MaStR/issues/202 and https://github.com/OpenEnergyPlatform/open-MaStR/issues/203

This feature will allow downloading the daily [Gesamtdatenauszug](https://www.marktstammdatenregister.de/MaStR/Datendownload) in a sqlite DB. 
The existing soap_API will be able to update this database to avoid downloading the Gesamtdatenauszug on a daily basis.

Ludee committed:
Include initial version of the xml parser #202

FlorianK13 committed:
Add files via upload

Bulk xml download, parser, writing to PostgreSQL

FlorianK13 committed:
Added skeleton version of Mastr class

FlorianK13 committed:
restructuring: xml_parser folder within open_mastr

FlorianK13 committed:
restructuring of xml_parser folder

FlorianK13 committed:
Mastr class init defined variables

FlorianK13 committed:
First changes to download class

FlorianK13 committed:
started xml to sqlite  code

FlorianK13 committed:
expanded the xml to sqlite utils function

FlorianK13 committed:
Changed to pep8 naming

FlorianK13 committed:
Added test mastr file

FlorianK13 committed:
created function add_table_to_sqlite_database

FlorianK13 committed:
worked on download bulk function

FlorianK13 committed:
Changed variables names to pep8 style guide

FlorianK13 committed:
added include_tables variable

FlorianK13 committed:
first stable to sqllite function

FlorianK13 committed:
deleted the meta data correction for sqlite database

FlorianK13 committed:
started the cleansing_sqlite_databse function

FlorianK13 committed:
restructuring

FlorianK13 committed:
refactoring the utils.py into three utils subfiles

FlorianK13 committed:
Started implementing the catalogue replacement in the bulk data cleansing

FlorianK13 committed:
worked on catalog cleansing for bulk download

FlorianK13 committed:
Complete catalogue replace code

FlorianK13 committed:
Create new parameter date_string in Mastr

When initializing a Mastr object, one can now give a date_string.
By using a date, where the zipped xml file was already downloaded and
lies on the local .open-MaStR/data/xml_download one avoids to redownload
the files.

[Ticket: 202]

FlorianK13 committed:
Change from postgres Error Exception to SQLite

FlorianK13 committed:
Apply python black to reformat code

FlorianK13 committed:
Correct minor bugs to avoid error raises

FlorianK13 committed:
Create new cleansed database entries

The original data is kept in the database. The tables where
entries are modified by replacing the values
from "katalogwerte.xml" are saved within new tables called
_cleansed.

FlorianK13 committed:
Create cleansing parameter

chrwm committed:
Update python version in environment.yml

chrwm committed:
Add script to mirror mastr

Use this script to mirror the mastr database and perform regular updates.

chrwm committed:
Add NetzbetreiberMastrNummer to class Extended

chrwm committed:
Add Registrierungsdatum to class Extended

FlorianK13 committed:
Add function annotation

FlorianK13 committed:
Merge branch 'feature-#202-xmlparser' of github.com:OpenEnergyPlatform/open-MaStR into feature-#202-xmlparser

FlorianK13 committed:
Add function annotation

FlorianK13 committed:
Add function annotation

FlorianK13 committed:
Create function for selecting tables

This function is needed, if only a subset of the existing tables should be written to the database.

FlorianK13 committed:
Add function annotation

FlorianK13 committed:
Delete parameter exclude_tables

The parameter exclude_tables is deleted from all the utils functions. In the download function,
the exclude_tables list is transferred to an inlcude_tables list.

FlorianK13 committed:
Create test functions

FlorianK13 committed:
Create test functions

FlorianK13 committed:
Reformat code with python black

FlorianK13 committed:
Replace katalogwerte in cleansed database

In the raw data, many entries are only given as numbers. Their meaning
can be read from the xml file katalogwerte. In the cleansed data, these
numbers are replaced by their original meaning. The table katalogkategorien
is not used anymore. Instead, a hardcoded list "columns_to_replace.py" is
created with the names of all columns were one has to replace the values.

FlorianK13 committed:
Reformat with python black

FlorianK13 committed:
Include all tables in full_list_of_files

FlorianK13 committed:
Delete entries that crashed the parsing

FlorianK13 committed:
Change folder name for better structure

FlorianK13 committed:
Create test for utils module

FlorianK13 committed:
Change order: First download xml file before using it

FlorianK13 committed:
Create test for write_sqlite module

FlorianK13 committed:
Add function description

FlorianK13 committed:
Create test for utils_write_sqlite module

FlorianK13 committed:
Update test functions for xml download

chrwm committed:
Add additional table columns to orm.py

The added table columns per class are queried by the API and require a column to be filled in.

chrwm committed:
Modify function flatten_dict

Adjust the function flatten_dict to solve an IndexError.

The function raised an 'IndexError: list index out of range', if a data point was processed that did not contain any 'VerknuepfteEinheiten'. The function tried to get the first, in this case non-existent, MaStRNummer and assign it as a new value to 'VerknuepfteEinheiten' (flatten dict). This led to the error.

The function was modified to check for empty lists for the key 'VerknuepfteEinheiten' and pass over an empty list in case of emptiness. Hence, the data point will have no 'VerknuepfteEinheiten'.

FlorianK13 committed:
Change type of parameter katalogwerte

FlorianK13 committed:
Delete unnecessary imports

chrwm committed:
Add mirror_mastr_csv_export.py

chrwm committed:
Add mirror_mastr_dump.py

chrwm committed:
Add mirror_mastr_processing_pipeline.py

FlorianK13 committed:
Change code to fulfill flake8

FlorianK13 committed:
Change code to fulfill flake8

FlorianK13 committed:
Change code to fulfill flake8

FlorianK13 committed:
Alter variable that is tested

FlorianK13 committed:
Change code to fulfill flake8

chrwm committed:
Add mastr-download.py

Use the script to switch between 'bulk' and 'API' download and to configure soap_API download settings.

chrwm committed:
Include 'API' download function

chrwm committed:
Modify function db_engine()

Remove hard-coded database URL and relocate it to settings.py to enable switching between postgresql and sqllite database and to increase user-friendliness.

chrwm committed:
Add settings.py

Use file to specify type and path of DB.
The path is specified in the file itself. The DB type is specified in the environment variables. Set variable "DB_ENGINE" either to "docker" for a postegresql DB in a docker container or "sqlite" for a sqlite DB.

chrwm committed:
Adapt to new structure of specifying DB type

chrwm committed:
Rename initdb() and adapt to changed DB setup

chrwm committed:
Add todos for additional required packages

FlorianK13 committed:
Add vscode to gitignore

FlorianK13 committed:
Add TODO

FlorianK13 committed:
Delete file that is not used in current implementation

FlorianK13 committed:
Add TODO

FlorianK13 committed:
Run python black on file

FlorianK13 committed:
Add TODO

FlorianK13 committed:
Correct the TODO syntax

FlorianK13 committed:
Add pytest fixture decorator

With the decorator, the connection con does not need to be executed in each function, but it is passed as an argument.
scope="module" makes the con function run only once at the beginning.
After the con parameter is yielded, the connection is closed in the second function call (at the end).

[Ticket: #202]

FlorianK13 committed:
Make file flake8 compatible

chrwm committed:
Revert "First draft of MaStR mirror"

This reverts commit 7a236cc374dff1276babdbca191b9cf5f352c8f2.

# Conflicts:
#	open_mastr/soap_api/download.py

chrwm committed:
Merge remote-tracking branch 'origin/feature-#202-xmlparser' into feature-#202-xmlparser

chrwm committed:
Add todos for further code development

FlorianK13 committed:
Change name of imported function _flatten_dict

FlorianK13 committed:
Subject line (try to keep under 50 characters)

Multi-line description of commit,
feel free to be detailed.

[Ticket: X]⌊倠敬獡⁥湥整⁲桴⁥潣浭瑩洠獥慳敧映牯礠畯⁲档湡敧⹳䰠湩獥猠慴瑲湩੧‣楷桴✠✣眠汩⁬敢椠湧牯摥‬湡⁤湡攠灭祴洠獥慳敧愠潢瑲⁳桴⁥潣浭瑩ਮਣ‣湏戠慲据⁨敦瑡牵ⵥ㈣㈰砭汭慰獲牥⌊夠畯⁲牢湡档椠⁳灵琠⁯慤整眠瑩⁨漧楲楧⽮敦瑡牵ⵥ㈣㈰砭汭慰獲牥⸧⌊⌊䌠慨杮獥琠⁯敢挠浯業瑴摥਺ण潭楤楦摥›†灯湥浟獡牴猯慯彰灡⽩業牲牯瀮੹ण潭楤楦摥›†灯湥浟獡牴猯慯彰灡⽩牯⹭祰⌊⌊䌠慨杮獥渠瑯猠慴敧⁤潦⁲潣浭瑩਺ण潭楤楦摥›†灯湥浟獡牴猯瑥楴杮⹳祰⌊

FlorianK13 committed:
Create a new way to handle the initialization of an sqlite database

sqlite does not support the CREATE SCHEMA command. Hence this can only be called if the database is
the postgres image. The mirror_schema is set to NONE for sqlite and the JSONB that can only be used
in postgres is set to a string in sqlite.

[Ticket: #202]

FlorianK13 committed:
Set sqlite database path

FlorianK13 committed:
Change variable name to reference use in bulk download

FlorianK13 committed:
Move url variable into utils function

FlorianK13 committed:
Move url variable into utils function

FlorianK13 committed:
Correct code to match flake8

FlorianK13 committed:
Refactor function initialize_db as method from Mastr class

The method to initialize the structure of the database has to be called
for both the API and the bulk download. It was therefore moved to a method
of the Mastr class. It is now called when a Mastr object is initialized.
Additionally, python black was used for code formatting.

[Ticket: #202]

FlorianK13 committed:
Rename variable to reference usage in bulk download

FlorianK13 committed:
Refactor variable to be defined only once

FlorianK13 committed:
Change database name

FlorianK13 committed:
Adapt to flake8

FlorianK13 committed:
Create sql engine and connection as clas attribute

FlorianK13 committed:
Delete old xml parser

FlorianK13 committed:
Adapt ORM for bulk download

The bulk download adds new tables, which are added to orm.py
Those tables have only their name and the primary key defined. A description of
all columns is added in future.
Second, the dictionary tablename_mapping is introduced.
It maps the table names from the bulk download to the names from the API
download.

[Ticket: 202]

FlorianK13 committed:
Add TODO

FlorianK13 committed:
Refactor tablename_mapping to orm.py

The bulk download is now written to predefined sql tables
with a predefined data model.

[Ticket: #202]

FlorianK13 committed:
Move database connection info to class attribute

FlorianK13 committed:
Rename Database path variable

chrwm committed:
Fix function name

FlorianK13 committed:
Try to change data types when writing to sqlite

FlorianK13 committed:
Add __class__ key and run python black

FlorianK13 committed:
Change Pandas requirements

dtype Int64 needs pandas 1.4

[Ticket: #202]

FlorianK13 committed:
Change source for katalogwerte

Katalogwerte should not be part of the final sql database.
Hence it is not written to sqlite anymore. The info is
drawn directly from the zipped xml file.

[Ticket: #202]

FlorianK13 committed:
Update StorageUnit and BalancingArea class

FlorianK13 committed:
Add print message for cleansing

FlorianK13 committed:
Correct table mapping name

FlorianK13 committed:
Correct sqlite writing bug

If the table is written again to sql, it first has to
be deleted (otherwise the not unique constraint does not allow
to add new elements).
Additionally, the function write_single_entries_until_not_unique_comes_up
was created, since there appear entries with the same primary key in the
xml data. These entries are simply deleted.

[Ticket: #202]

FlorianK13 committed:
Changed datatype of gemeindeschluessel

FlorianK13 committed:
Add 0 at start of PLZ and Gemeindeschluessel

PLZ and Gemeindeschluessel are a 5 and 8 digit string that
can start with a 0. When they are parsed, they are parsed
as a float. So the PLZ of 01234 becomes 1234. With the new
function, a zero is added to all PLZ and Gemeindeschluessel
where it is missing.

[Ticket: #202]

FlorianK13 committed:
Replace values from system_catalogue

In the documentation of the marktstammdatenregister gesamtdatenauszug
(https://www.marktstammdatenregister.de/MaStRHilfe/files/gesamtdatenexport/Dokumentation%20MaStR%20Gesamtdatenexport.pdf)
they write about the Systemkatalog. This is a replacement for some values in a few
columns. Those integer values are replaced now with their correspanding string value.

[Ticket: #202]

FlorianK13 committed:
Move orm.py to main higher folder

FlorianK13 committed:
Update test to new class attributes

FlorianK13 committed:
Add data model for Grids

FlorianK13 committed:
Add sqlite3 exception for missing column

FlorianK13 committed:
Update documentation of mastr class for release 1.0.0

FlorianK13 committed:
Update mastr class + bulk download description

chrwm committed:
Revert "Revert "First draft of MaStR mirror""

This reverts commit b4888947a1d97ba2503bd948b8cc42d4df12e36b.

chrwm committed:
Fix function naming

For functions flatten_dict and basic_unit_data after revert b4999466ea264c4566e388b6fc1baabcb3925ee5

FlorianK13 committed:
Update requirements and add comments

FlorianK13 committed:
Add descriptions for variables

FlorianK13 committed:
Add import of Mastr class

FlorianK13 committed:
Add Exception for wrong method variable

FlorianK13 committed:
Update documentation

FlorianK13 committed:
Start parameter description of download method

FlorianK13 committed:
Highlight important text

FlorianK13 committed:
Change file type to svg

FlorianK13 committed:
Add more details to bulk download description

FlorianK13 committed:
Change data type of column DatumLetzteAktualisierung

chrwm committed:
Add optional data type for columns

Switch between data type, to ensure compatability with either postgresql or sqlite database

FlorianK13 committed:
Replace data type string with date

FlorianK13 committed:
Merge branch 'feature-#202-xmlparser' of github.com:OpenEnergyPlatform/open-MaStR into feature-#202-xmlparser

chrwm committed:
Adapt data types to match column data types from orm

chrwm committed:
Uncomment soap_API download code

chrwm committed:
Format with black

deniztepe committed:
Set default for invalid datetime entries

Some invalid datetime strings raise error when `pd.datetime()` is applied. With the 'coerce' option, the invalid strings such as '2022-01-32' or '202-01-01' are converted to NaT

deniztepe committed:
Extend prompt messages and correct typos

Prompt messages for successful bulk download and cleansing are added. Black is applied. Some english typos are corrected.

deniztepe committed:
Add exception for invalid bulk_date_string

Only current MaStR file can be downloaded from the website.
If a previous date is given and the .zip file does not exist, the function `mastr.download()` should raise an exception instead of downloading the up to date MaStR file.

FlorianK13 committed:
Delete variable bulk_exclude_tables

This variable could be used to name tables that should not be included. There is no
real application scenario that anyone will use this variable. It is therefore deleted
to make the function more accessible.

[Ticket: #202]

FlorianK13 committed:
Extend docstring

deniztepe committed:
Use the parameter technology instead of bulk_include_tables

The selection of tables to be created/filled happens for bulk download also by technology. With a further function the technologies are mapped to the .xml file names.
This way the parameter of mastr.download() 'bulk_include_tables' is avoided and the function usage is standardized for bulk and API.

deniztepe committed:
Correct data type of a column

deniztepe committed:
Avoid unnecessary parameter to overwrite raw data tables

deniztepe committed:
Match the datatype of tables via bulk download with the orm structure

When the dataframe is written in the database, the datatypes in the orm structure are being respected. This step must be done in the first converting as well as after cleansing.

deniztepe committed:
Perform bulk cleansing before writing into sqlite database

In order to avoid second interaction with the database for cleansing, the procedure will be performed right after converting the xml file and before writing the table into the database. Additionally, the option for conversion to datetime datatypes is included into the bulk cleansing.
Black is applied

deniztepe committed:
Avoid setup of docker if  'sqlite' is chosen

Unnecessary setup of docker container was caused by initialiye_db. The parameter is changed to DB_ENGINE, so that the function 'setup_docker' is called only if docker is chosen.

deniztepe committed:
Expand filter for NetzbetreiberMastrNummer

FlorianK13 committed:
Closes #213, Add Column for data source

In each table that can be filled by both API and bulk download, we now see a new column named "Quelle" (german for source).
It has the two possible entries "API" and "bulk".

[Ticket: #213]

FlorianK13 committed:
Check if location_type is not null before iterating

FlorianK13 committed:
Change printed phrase for cleansing

FlorianK13 committed:
Add column for katalogeintraege replacement

FlorianK13 committed:
Add data description for bulk data

Some tables are only filled by the bulk download. For those tables, the data definition was added to the orm.py file.

[Ticket: #209]

FlorianK13 committed:
Change data types in orm.py classes

FlorianK13 committed:
Add missing columns for bulk download

These missing columns seem to be responsible for a bug that appeared when writing data to the database a second time
with the bulk download.

FlorianK13 committed:
Change column name to DatenQuelle

FlorianK13 committed:
Add column DatumDownload for bulk download

When data is downloaded via bulk and written to the database, a column DatumDownload exists within every table.
This column is now filled with the date, when the zipped xml files were downloaded.

deniztepe committed:
Add columns originated from bulk into orm

FlorianK13 committed:
Rename parameters that are API related

deniztepe committed:
Change expected datatype of bulk_download_date

deniztepe committed:
Map column name from bulk to API format

The column 'Verknüpfte Einheiten' is named differently in bulk and API method, but consists of same information. The column name is mapped to 'VerknuepfteEinheit'

deniztepe committed:
Black applied

deniztepe committed:
Remove entirely empty columns from orm structure

FlorianK13 committed:
Update documentation

FlorianK13 committed:
Correct spelling mistake

deniztepe committed:
Add columns originated from bulk into orm

Follow up commit of 413720e

deniztepe committed:
Rename parameters with api prefix

TODO comment is deleted

FlorianK13 committed:
Add docstring to download method

deniztepe committed:
Revert "Remove entirely empty columns from orm structure"

This reverts commit 7407ac45dd1e41c9b30eb71beb2277fd26b68983.
API download does not work without the previously removed columns, even if they remain empty.

FlorianK13 committed:
Change data type of DatumDownload

FlorianK13 committed:
Add DatumDownload to API download

FlorianK13 committed:
Add new possible value to parallel_process

On Windows, one gets trapped in an infinite loop for download(method="API", api_processes=None). Hence
a new value "max" is introduced. Only with api_processes="max", the maximum possible value for api_processes
is determined and used. Otherwise no parallel process is started.

[Ticket: #211]

FlorianK13 committed:
Update changes in Changelog

FlorianK13 committed:
Delete TODOs

FlorianK13 committed:
Add column to be cleansed

FlorianK13 committed:
Update docstrings

FlorianK13 committed:
Update Documentation

FlorianK13 committed:
Create function for parameter validity check

FlorianK13 committed:
Create test for parameter validity check

In the two new test functions, both working and none working parameters are checked.

[Ticket: #207]

deniztepe committed:
Clean expired functions and parameters

deniztepe committed:
Clean expired parameter

Script has been shortened without changing the syntax

deniztepe committed:
Clean expired functions

FlorianK13 committed:
Change Raised error type and add technology validity check

The raised error is now a ValueError. Additionally, the validity of parameter technologies is checked in the _validate_parameter_format_for_download_method

FlorianK13 committed:
Merge branch 'feature-#202-xmlparser' of github.com:OpenEnergyPlatform/open-MaStR into feature-#202-xmlparser

FlorianK13 committed:
Add test for parameter validity check

In the first function, a combination of parameters is tested that should not cause an error.
In the second test function, combinations of parameters are tested that should cause an error.

[Ticket: #207]

FlorianK13 committed:
Resolve error caused by parameter=None in download method

Some parameters in the download method are initialized as None. They must be allowed to be None
within the parameter validity check.

[Ticket: #207]

deniztepe committed:
Update and adapt tests

The unit tests are adapted to the changed function structure. Some tests are skipped whit pytest.mark.skipif() , if the data files are not existing.

[Ticket: #207]

deniztepe committed:
Correct typo in docstring

deniztepe committed:
Add tests for utils_write_sqlite

A test database will be created and removed immedeately.

deniztepe committed:
Rename table_name to column_name

deniztepe committed:
Add further tests for bulk download functions

* Remove depricated tests
* Silence a warning because of clint module
* Rename testdb_con -> con_testdb to avoid a misunderstanding of fixture as test

[Ticket: #207]

FlorianK13 committed:
Change input of parameter validity check

The former input of one dictionary is not useful. Now the parameter
validity check function gets the same parameters as the download function.

[Ticket: #207]

FlorianK13 committed:
Add warning for ignored parameters

For method=bulk, api related parameters are ignored by the algorithm
and vice versa.
This is now shown to the user by a warning.

deniztepe committed:
﻿Refactor: Move function from orm to mastr

deniztepe committed:
Cancel optional conversion to datetime datatypes

to_datetime() is pulled out of bulk_cleansing option, since without the conversion the datatypes don't match with these defined in orm.py

FlorianK13 committed:
Change parameter type from str to list

The two parameters api_data_types and api_location_types have to be
lists. Hence the tests and the validity check function were adapted.
Also, black was run on the code.

chrwm committed:
Fix typo in ci.yml

On GitHub Actions the secrets are saved spelled: "MASTR_TOKEN" and "MASTR_USER". The typo in ci.yml "MASTR_User" leads to the error of the var user being null.

(cherry picked from commit c5dcd5bd50bf1e647aac2e7ffe3d52b4117944cd)

deniztepe committed:
Allow for download parameters to be None

api_location_types and api_data_types were not allowed to be 'None'. The validation function is adapted and the corresponding test case is added.

deniztepe committed:
Update mastr_mirror tests

Tests are now up to date and should be running successfully at that moment.
Black is applied.

[Ticket: 207]

deniztepe committed:
Initial commit for ci-test

Try to run real pytest tests on repository with github ci

deniztepe committed:
Set windows-2019

deniztepe committed:
Try without windows

deniztepe committed:
Add missing requirements

deniztepe committed:
Try to install requirements correctly

FlorianK13 committed:
Restructure documentation

There are now two main sections on the documentation website:
Getting started and Advanced. Detailed description of the package went
into the Advanced section.

FlorianK13 committed:
Update requirements

Update requirements in the setup.py file. In future, the requirements.txt
can be deleted. Additionally, python versions were changed
to >3.7 and black was run.

FlorianK13 committed:
Update documentation

Since the dependencies are now stated in setup.py, this command changes.

curiousleo commented:
This may be of interest, if perhaps only as a point of comparison: with https://github.com/marktstammdatenregister-dev/mastr/pull/31, the current Gesamtdatenauszug / Gesamtdatenexport can be downloaded as a gzipped SQLite file from https://s3.eu-central-1.wasabisys.com/mastr-backup/Marktstammdatenregister.db.gz.

```console
$ curl https://s3.eu-central-1.wasabisys.com/mastr-backup/Marktstammdatenregister.db.gz | gunzip - >Markstammdatenregister.db
$ sqlite3 --readonly Markstammdatenregister.db 'select round(sum(Nettonennleistung), 3) from EinheitSolar'
57509349.779
```

The same SQLite database can be browsed as a Datasette instance on https://ds.marktstammdatenregister.dev/Marktstammdatenregister.

FlorianK13 committed:
Delete wrong whitespace

FlorianK13 committed:
Add missing dependencies

deniztepe committed:
Store raw data information in git comform utf-8 csv format

Created csv files are now in utf-8 format. Sphinx run to update the documentation.

deniztepe committed:
Add matrix strategy for tests and remove docker process from ci

bash chosen since if is not known by windows terminal

deniztepe committed:
Change from requirements.txt to setup.py

deniztepe committed:
Merge branch 'ci-test' into feature-#202-xmlparser

deniztepe committed:
Change branch name at ci workflow

deniztepe committed:
Rename ci workflow and change the trigger to the push on dev branch

chrwm committed:
Update documentation

chrwm committed:
Delete download script

chrwm committed:
Delete unnecessary print statements

FlorianK13 committed:
Delete python 3.7 as programming language

Python 3.7 does not support pandas 1.4, but pandas 1.4 is
needed for the read_xml command.

FlorianK13 committed:
Change data types

FlorianK13 committed:
Add to_csv function

The function is used to read existing sqlite database and
write the tables into single csv files.

[Ticket: #219]

FlorianK13 committed:
Delete ontology folder

The ontology is not developed within open_mastr, hence
those files are deleted.

FlorianK13 committed:
Relocate Folder postprocessing

Postprocessing files are mainly jupyter notebooks that can be used to work
with the package. But they should not be included in the package, so they
are removed from the folder open_mastr.

[Ticket: #205]

FlorianK13 committed:
Delete print command

The print command was executed when importing the package and is now
deleted.

FlorianK13 committed:
Apply flake8

FlorianK13 committed:
Extend documentation in README

deniztepe committed:
Add missing dependency to install_requires

A fresh import to an empty environment has lead to an import error

deniztepe committed:
Reorder the imports

An initial installation of the package depends on this order. The error occured also at remote ci workflow.
The warning from flake8 is ignored by an inline comment

deniztepe committed:
Add the Mastr class documentation the references in readthedocs

FlorianK13 committed:
Add data model of MaStR

The data model is added to the documentation, together with its source.

FlorianK13 committed:
Apply flake8

FlorianK13 commented:
Temporarily closed to save github resources (with an open pull request, tests are run for every commit)
